src/openrct2-ui/windows/EditorObjectiveOptions.cpp

```diff
@@ -191,925 +132,1019 @@ static uint64_t window_editor_objective_options_page_hold_down_widgets[] = {
-    w->hold_down_widgets = window_editor_objective_options_page_hold_down_widgets[WINDOW_EDITOR_OBJECTIVE_OPTIONS_PAGE_MAIN];
-    WindowInitScrollWidgets(*w);
-    w->selected_tab = WINDOW_EDITOR_OBJECTIVE_OPTIONS_PAGE_MAIN;
-    w->no_list_items = 0;
-    w->selected_list_item = -1;
-    WindowEditorObjectiveOptionsUpdateDisabledWidgets(w);
-
-    return w;
-}
-
-static void WindowEditorObjectiveOptionsSetPressedTab(WindowBase* w)
-{
-    int32_t i;
-    for (i = WIDX_TAB_1; i <= WIDX_TAB_2; i++)
-        w->pressed_widgets &= ~(1 << i);
-    w->pressed_widgets |= 1LL << (WIDX_TAB_1 + w->page);
-}
-
-static void WindowEditorObjectiveOptionsAnchorBorderWidgets(WindowBase* w)
-{
-    w->ResizeFrameWithPage();
-}
-
-static void WindowEditorObjectiveOptionsDrawTabImages(WindowBase* w, DrawPixelInfo* dpi)
+class EditorObjectiveOptionsWindow final : public Window
 {
-    Widget* widget;
-    int32_t spriteIndex;
-
-    // Tab 1
-    widget = &w->widgets[WIDX_TAB_1];
-
-    spriteIndex = SPR_TAB_OBJECTIVE_0;
-    if (w->page == WINDOW_EDITOR_OBJECTIVE_OPTIONS_PAGE_MAIN)
-        spriteIndex += (w->frame_no / 4) % 16;
-
-    GfxDrawSprite(dpi, ImageId(spriteIndex), w->windowPos + ScreenCoordsXY{ widget->left, widget->top });
-
-    // Tab 2
-    if (!WidgetIsDisabled(*w, WIDX_TAB_2))
+public:
+    void OnOpen() override
     {
-        widget = &w->widgets[WIDX_TAB_2];
-        spriteIndex = SPR_TAB_RIDE_0;
-        if (w->page == WINDOW_EDITOR_OBJECTIVE_OPTIONS_PAGE_RIDES)
-            spriteIndex += (w->frame_no / 4) % 16;
-
-        GfxDrawSprite(dpi, ImageId(spriteIndex), w->windowPos + ScreenCoordsXY{ widget->left, widget->top });
+        widgets = window_editor_objective_options_main_widgets;
+        pressed_widgets = 0;
+        hold_down_widgets = window_editor_objective_options_page_hold_down_widgets[WINDOW_EDITOR_OBJECTIVE_OPTIONS_PAGE_MAIN];
+        InitScrollWidgets();
+        selected_tab = WINDOW_EDITOR_OBJECTIVE_OPTIONS_PAGE_MAIN;
+        no_list_items = 0;
+        selected_list_item = -1;
+        UpdateDisabledWidgets();
     }
-}
 
-/**
- *
- *  rct2: 0x00668496
- */
-static void WindowEditorObjectiveOptionsSetPage(WindowBase* w, int32_t page)
-{
-    if (w->page == page)
-        return;
-
-    w->page = page;
-    w->frame_no = 0;
-    w->var_492 = 0;
-    w->no_list_items = 0;
-    w->selected_list_item = -1;
-    w->hold_down_widgets = window_editor_objective_options_page_hold_down_widgets[page];
-    w->event_handlers = window_editor_objective_options_page_events[page];
-    w->widgets = window_editor_objective_options_widgets[page];
-    w->Invalidate();
-    WindowEditorObjectiveOptionsUpdateDisabledWidgets(w);
-    WindowEventResizeCall(w);
-    WindowEventInvalidateCall(w);
-    WindowInitScrollWidgets(*w);
-    w->Invalidate();
-}
+    void OnMouseUp(WidgetIndex widgetIndex) override
+    {
+        switch (widgetIndex)
+        {
+            case WIDX_CLOSE:
+                Close();
+                break;
```

Comment:
return for this and l160. This is important as after close the class is invalid so we don't want anything else to be accessed.

Reply:
My bad, yeah I remember having to add something like this to my previous PR too - will include in the next push!